dot/parachain/backing/active_leaves_update.go

```diff
@@ -0,0 +1,528 @@
+		coreValue, err := cores.Types[idx].Value()
+		if err != nil {
+			return nil, fmt.Errorf("getting core value at index %d: %w", idx, err)
+		}
+
+		var coreParaID parachaintypes.ParaID
+		switch v := coreValue.(type) {
+		case parachaintypes.OccupiedCore:
+			if mode.IsEnabled {
+				// Async backing makes it legal to build on top of occupied core.
+				coreParaID = parachaintypes.ParaID(v.CandidateDescriptor.ParaID)
+			} else {
+				continue
+			}
+		case parachaintypes.ScheduledCore:
+			coreParaID = v.ParaID
+		case parachaintypes.Free:
+			continue
+		}
+
+		coreIndex := parachaintypes.CoreIndex{Index: uint32(idx)}
+		groupIndex := validatorGroups.GroupRotationInfo.GroupForCore(coreIndex, numOfCores)
+		validatorIndexes := validatorGroups.Validators[groupIndex]
+
+		if validatorIndexes != nil {
+			if localValidator != nil && slices.Contains(validatorIndexes, localValidator.index) {
+				assignment = &coreParaID
+			}
+			groups[coreParaID] = validatorIndexes
+		}
+	}
+
+	tableContext := TableContext{
+		validator:  localValidator,
+		groups:     groups,
+		validators: validators,
+	}
+
+	tableConfig := Config{
+		AllowMultipleSeconded: mode.IsEnabled,
+	}
+
+	newPerRelayParentState := perRelayParentState{
+		prospectiveParachainsMode: mode,
+		relayParent:               relayParent,
+		assignment:                assignment,
+		table:                     newTable(tableConfig),
+		tableContext:              tableContext,
+		fallbacks:                 make(map[parachaintypes.CandidateHash]attestingData),
+		awaitingValidation:        make(map[parachaintypes.CandidateHash]bool),
+		issuedStatements:          make(map[parachaintypes.CandidateHash]bool),
+		backed:                    make(map[parachaintypes.CandidateHash]bool),
+		minBackingVotes:           minBackingVotes,
+	}
+
+	return &newPerRelayParentState, nil
+}
+
+func minBackingVotes(rt runtime.Instance) (uint32, error) {
+	votes, err := rt.ParachainHostMinimumBackingVotes()
+	if err != nil && errors.Is(err, wazero_runtime.ErrExportFunctionNotFound) {
+		logger.Tracef(
+			"%s is not supported by the current Runtime API",
+			runtime.ParachainHostMinimumBackingVotes,
+		)
+		return LEGACY_MIN_BACKING_VOTES, nil
+	}
+	return votes, err
+}
+
+func fetchParachainHostData(rt runtime.Instance) (
+	*parachaintypes.SessionIndex,
+	[]parachaintypes.ValidatorID,
+	*parachaintypes.ValidatorGroups,
+	*scale.VaryingDataTypeSlice,
+	error,
+) {
+	var (
+		sessionIndex    parachaintypes.SessionIndex
+		validators      []parachaintypes.ValidatorID
+		validatorGroups *parachaintypes.ValidatorGroups
+		cores           *scale.VaryingDataTypeSlice
+	)
+
+	// Create a context with cancellation capability.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel() // Ensure cancellation happens when function returns.
+
+	// WaitGroup to wait for all goroutines to finish.
+	var wg sync.WaitGroup
+
+	// Error channel to receive errors from goroutines.
+	errCh := make(chan error)
```

Comment:
```suggestion
	errCh := make(chan error, 4)
```

let's use a buffered channel and read all the errors in the end at the same time.